Sources/SwiftKafka/KafkaPollingSystem.swift

```diff
@@ -0,0 +1,299 @@
+        return KafkaAsyncIterator(wrappedIterator: self.wrappedSequence.makeAsyncIterator())
+    }
+}
+
+/// A back-pressure aware polling system for managing the poll loop that polls `librdkafka` for new acknowledgements.
+final class KafkaPollingSystem<
+    Element,
+    BackPressure: NIOAsyncSequenceProducerBackPressureStrategy
+>: Sendable {
+    /// The producer type used in the system.
+    typealias Producer = NIOAsyncSequenceProducer<Element, BackPressure, KafkaPollingSystem>
+
+    /// The state machine that manages the system's state transitions.
+    let stateMachineLock: NIOLockedValueBox<StateMachine>
+
+    /// Closure that takes care of polling `librdkafka` for new messages.
+    var pollClosure: (() -> Void)? {
+        get {
+            self.stateMachineLock.withLockedValue { stateMachine in
+                return stateMachine.pollClosure
+            }
+        }
+        set {
+            self.stateMachineLock.withLockedValue { stateMachine in
+                stateMachine.pollClosure = newValue
+            }
+        }
+    }
+
+    /// Initializes the ``KafkaBackPressurePollingSystem``.
+    /// Private initializer. The ``KafkaBackPressurePollingSystem`` is not supposed to be initialized directly.
+    /// It must rather be initialized using the ``KafkaBackPressurePollingSystem.createSystemAndSequence`` function.
+    /* USE createSystemAndSequence, TESTING only */ init() {
+        self.stateMachineLock = NIOLockedValueBox(StateMachine())
+    }
+
+    /// Factory method creating a ``KafkaBackPressurePollingSystem`` and the ``AsyncSequence`` that receives its messages .
+    /// The caller of this function must retain the sequence in order to receive messages.
+    ///
+    /// - Returns: A tuple containing the ``KafkaBackPressurePollingSystem`` and a reference to the ``KafkaAsyncSequence``.
+    static func createSystemAndSequence(
+        backPressureStrategy: BackPressure
+    ) -> (KafkaPollingSystem, KafkaAsyncSequence<Element, BackPressure>) {
+        let pollingSystem = KafkaPollingSystem()
+
+        // (NIOAsyncSequenceProducer.makeSequence Documentation Excerpt)
+        // This method returns a struct containing a NIOAsyncSequenceProducer.Source and a NIOAsyncSequenceProducer.
+        // The source MUST be held by the caller and used to signal new elements or finish.
+        // The sequence MUST be passed to the actual consumer and MUST NOT be held by the caller.
+        // This is due to the fact that deiniting the sequence is used as part of a trigger to
+        // terminate the underlying source.
+        let acknowledgementsSourceAndSequence = NIOAsyncSequenceProducer.makeSequence(
+            elementType: Element.self,
+            backPressureStrategy: backPressureStrategy,
+            delegate: pollingSystem
+        )
+
+        pollingSystem.stateMachineLock.withLockedValue { stateMachine in
+            stateMachine.sequenceSource = acknowledgementsSourceAndSequence.source
+        }
```

Comment:
Can we rather pass this in the init of the `KafkaPollingSystem`?

Reply:
We cannot pass everything on `init` as we have a lot of circular dependencies, e.g. `NIOAsyncSequenceProducer` needs `KafkaPollingSystem` as its delegate, `KafkaPollingSystem` needs `NIOAsyncSequenceProducer.Source` to produce new elements. So at some point, we have to assign the references after initialization.